Added Config class and project readMe

hugofdevoteam opened the pull request.

Commit:
Added Config class and project readMe

Commit:
Update package name

Commit:
Update version of codehaus on pom

Commit:
Update version of codehaus on pom

Commit:
Update multiple sources

Update project name on pom and readme
Add RequestSender migrated class

hugofdevoteam commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,215 @@
+package com.qudini.api;
```
This class has been only migrated, it has no changes except for the class imports and configurations that would break the build

Commit:
Update multiple sources

Add new config obj requestStaticValues to the config files
Update configuration class
Update naming of runner to QudiniAppConfDataGenerator
Update RequestSender with partial refactor of the class

Commit:
Update RequestSender with all the refactor needed

Commit:
Add Merchant class

Commit:
Update project with data and classes needed for merchant and venue handling

duartejoel commented on README.md:
```diff
@@ -1,2 +1,23 @@
+This project is heavily based on the code already produced to populate the qudini app, the main changes are refactor of 
+possible deprecated methods, addiction on basic logging instead of STDOUT and possible runner customizations.   
+
+The primary intend of project this to try to decouple this code from the test project and this way to act has a standalone
```
... of this project is decoupling ...

hugofdevoteam replied:
done. thx

duartejoel commented on pom.xml:
```diff
@@ -0,0 +1,92 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <encoding>UTF-8</encoding>
+    </properties>
+
```
Just a reminder that we will need to add scm repo config here.

duartejoel commented on src/main/java/com/qudini/QudiniAppDataGenerator.java:
```diff
@@ -0,0 +1,39 @@
+
+    public static void main(String args[]) throws IOException {
+
+//        Merchants merchants = new Merchants();
```
I assume this is a wip

hugofdevoteam replied:
Yes, the "runner" will be the last thing to be made, for now is just to "test" that runs ok

duartejoel commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,490 @@
+            String charSet,
+            boolean withCookies) {
+
+        HttpPost httpPost;
```
You can inline the init of this var here.

hugofdevoteam replied:
Done. Thx

duartejoel commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,490 @@
+
+            log.error(String.format("The provided URL [ %s ] does seem to be a valid URL", url));
+
+            throw new RuntimeException();
```
Consider creating an exception for this or adding a message to it.

hugofdevoteam replied:
The "message" is in the log.error. Should I add the same message for the Exception?

duartejoel replied:
Yes, that's good practise to give some context to thrown exceptions.

hugofdevoteam replied:
Ok,  will add it also to the exception 👍

duartejoel commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,490 @@
+
+        HttpPost httpPost;
+
+        if (url.startsWith("http:") || url.startsWith("https:")) {
```
Consider refactoring this if-else statement that is used in multiple methods almost in the same way.

duartejoel commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,490 @@
+
+    }
+
+    private HttpPut httpPutBaseSpecification(String url, String contentType, String token, String paramsAsString, String charSet) {
```
Is not this the same as httpPostBaseSpec?

hugofdevoteam replied:
In a nutshell yes you are right, my problem is that the instance to be created is a HttpPut instead of a HttpPost and also returns the same instance.
For that I probably need to use a generic, but I could find out how. 
This comment applies to the if-else refactor

duartejoel replied:
You could returning something like https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/methods/HttpRequestBase.html or https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/methods/HttpEntityEnclosingRequestBase.html which are super classes of HttpPost and HttpPut.

hugofdevoteam replied:
Ok will try to do this simplification then. Thank you to pointing a way.

duartejoel commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,490 @@
+
+    }
+
+    private Map<String, String> httpResponseDecomposer(CloseableHttpResponse httpResponse) throws IOException {
```
Consider creating a new Dto for this instead of using a generic Map. It will make other methods less complex and easier to read.

hugofdevoteam replied:
Ok will do that

duartejoel commented on src/main/java/com/qudini/api/requests/composition/Merchants.java:
```diff
@@ -0,0 +1,174 @@
+    public void createMerchant(List<NameValuePair> merchantNameValuePairs,
+                               String envBaseUri,
+                               String qudiniAppUsername,
+                               String qudiniAppPassword) throws UnsupportedEncodingException {
```
What do you think of using the token instead of (user, pass) and include it only in the request sender where you are adding the headers? You could also include the token in the config instead of (user, pass).

hugofdevoteam replied:
> You could also include the token in the config instead of (user, pass).

It may be done, but the reasons behind I not doing that was:
- although token is static generated, it still needs to be generated (even if it is only one time for each user/pass), if I added it to the configuration we probably should have a "service" user to use is token instead of this admin user
- user/pass allow anyone to user his user to perform any experience and not being "locked" to some user/pass or token. Also it may be extended in the future to be able to get those from cmd line

> What do you think of using the token instead of (user, pass) and include it only in the request sender where you are adding the headers? 

I may not fully understand what you mean here, but if I understood correctly my intent was to not "access" the generate token directly from the "runner" (although not it is not done) so only the classes that would extend RequestSender would be able to generate the token. That is why all of them get the user/pass instead of the token.
Again if you see that may be "too much complication" for this type of data-generation I completely trust your judgement and will change the classes so the token is either generated in the runner and passed as parameter or already be placed on the config file.

duartejoel replied:
If the pair is always read from the config file I would read it directly in the RequestSender, generate the token there and pass it in every request. Consumers of RequestSender would not need to know about user/pass or token and no need to pass them in every method. Or have a RequestSender init method to init user/pass and baseuri as they don't change after being set.

hugofdevoteam replied:
_Yes seems a good idea, I will do that refactor. 
Thank you for pointing it out._

EDIT:
Strike what is said above.

> If the pair is always read from the config file I would read it directly in the RequestSender, generate the token there and pass it in every request.

For standalone run, yes it is read from config you have a point there, but when I have the jar as a lib I need to be able provide this "static" values.

>Or have a RequestSender init method to init user/pass and baseuri as they don't change after being set.

Seems the best solution for me, I will try to impl in a diff branch and then talk to you.

Commit:
Add queue creation classes

Add queue creation classes
Update project with some PR reviews

duartejoel commented on src/main/java/com/qudini/api/RequestSender.java:
```diff
@@ -0,0 +1,490 @@
+            String charSet) {
+
+        HttpPost httpPost;
+        httpPost = httpPostBaseSpecification(url, contentType, token, paramsAsString, charSet);
```
Could also be inlined here.

hugofdevoteam replied:
Done 👍

Commit:
Update venue resources

Commit:
Update request sender class with refactor

Commit:
Update request sender class with refactor

Commit:
Update request sender class with refactor of httpBaseSpecification

Commit:
Update classes wit PR refactor

Commit:
Update applied formatter to the code

Commit:
Add Products class

Add Products handling for queue and booking types
Add QudiniAppResponseDataUtils

Commit:
Add Booking Widget enable logic

Commit:
Update project start adding booking

Commit:
WIP - Update products to queues

Commit:
WIP - Fix association between products and queues

Commit:
Fix association between products and queues

Commit:
Add simple booking widget creation script

Add simple booking widget creation script
Update readMe with more information

Commit:
Fix corrections in readMe file

Commit:
Update project with IJ code formatter

Commit:
Add project work from thuesday

Commit:
Add remaining project
